File: apps/ios/MagicMobile/OnDeviceMessageLog.swift
import Foundation
import MagicMobileOnDevice

/// Retains only this authenticated seat's explicit informational events in memory.
/// Not a complete action log, durable game save, or source of hidden card details.
struct OnDeviceMessageLog {
    private var matchID: String?
    private var seatID: String?
    private var processedThrough: Int64 = -1
    private(set) var entries: [GameLogEntry] = []
    private var messageBytes = 0
    private static let retainedMessages = 64
    private static let retainedBytes = 1024 * 1024

    /// Validate the entire batch before changing state. Duplicate/older polls do
    /// not duplicate notices, and another match or seat can never reuse this log.
    mutating func ingest(_ poll: MatchPoll) throws {
        guard !poll.matchID.isEmpty, !poll.seatID.isEmpty,
              matchID == nil || matchID == poll.matchID,
              seatID == nil || seatID == poll.seatID else { throw EngineError.unboundPeer }
        guard poll.revision >= processedThrough else { return }
        let events: [MagicMobileOnDevice.JSONValue]
        if let supplied = poll.raw["events"] {
            guard let values = supplied.array, values.count <= 128 else {
                throw EngineError.invalidMessage("Malformed engine event history")
            }
            events = values
        } else { events = [] }
        var next = entries
        var bytes = messageBytes
        var previous: Int64 = -1
        for event in events {
            guard let revision = event["revision"]?.integer, revision >= 0,
                  revision <= poll.revision, revision > previous,
                  let kind = event["kind"]?.string else {
                throw EngineError.invalidMessage("Malformed engine event revision")
            }
            previous = revision
            guard revision > processedThrough, kind == "message" else { continue }
            guard let body = event["body"]?.object else {
                throw EngineError.invalidMessage("Malformed private engine notice")
            }
            // Upstream can emit an empty informational message. Do not invent text.
            if body["message"] == .null { continue }
            guard let message = body["message"]?.string else {
                throw EngineError.invalidMessage("Private engine notice has no text")
            }
            guard !message.isEmpty else { continue }
            next.append(GameLogEntry(id: "xmage:\(poll.matchID):\(poll.seatID):\(revision)",
                                     message: message, createdAt: nil))
            bytes += message.utf8.count
            // Keep the latest message intact, even when it alone exceeds the
            // history budget. The transport already bounds each entire JSON reply.
            while next.count > Self.retainedMessages || (bytes > Self.retainedBytes && next.count > 1) {
                bytes -= next.removeFirst().message.utf8.count
            }
        }
        matchID = poll.matchID; seatID = poll.seatID
        processedThrough = poll.revision; entries = next; messageBytes = bytes
    }
}

File: apps/android/core/src/main/kotlin/io/magicmobile/android/game/OnDeviceMessageLog.kt
package io.magicmobile.android.game

import kotlinx.serialization.json.JsonNull

/**
 * Port of apps/ios/MagicMobile/OnDeviceMessageLog.swift. Retains only this seat's explicit
 * informational events in memory. Not a complete action log or a source of hidden details.
 */
class OnDeviceMessageLog {
    private var matchID: String? = null
    private var seatID: String? = null
    private var processedThrough = -1L
    var entries: List<GameLogEntry> = emptyList(); private set
    private var messageBytes = 0

    fun copy(): OnDeviceMessageLog = OnDeviceMessageLog().also {
        it.matchID = matchID; it.seatID = seatID; it.processedThrough = processedThrough
        it.entries = entries; it.messageBytes = messageBytes
    }

    /** Validate the whole batch before changing state. Older polls never duplicate notices. */
    fun ingest(poll: MatchPoll) {
        if (poll.matchID.isEmpty() || poll.seatID.isEmpty() || (matchID != null && matchID != poll.matchID) ||
            (seatID != null && seatID != poll.seatID)) throw EngineError.UnboundPeer
        if (poll.revision < processedThrough) return
        val supplied = poll.raw["events"]
        val events = if (supplied != null) {
            val values = supplied.array
            if (values == null || values.size > 128) throw EngineError.InvalidMessage("Malformed engine event history")
            values
        } else emptyList()
        val next = entries.toMutableList()
        var bytes = messageBytes
        var previous = -1L
        for (event in events) {
            val revision = event["revision"].integer
            val kind = event["kind"].string
            if (revision == null || revision < 0 || revision > poll.revision || revision <= previous || kind == null) {
                throw EngineError.InvalidMessage("Malformed engine event revision")
            }
            previous = revision
            if (revision <= processedThrough || kind != "message") continue
            val body = event["body"].obj ?: throw EngineError.InvalidMessage("Malformed private engine notice")
            // Upstream can emit an empty informational message. Do not invent text.
            if (body["message"] is JsonNull) continue
            val message = body["message"].string ?: throw EngineError.InvalidMessage("Private engine notice has no text")
            if (message.isEmpty()) continue
            next += GameLogEntry("xmage:${poll.matchID}:${poll.seatID}:$revision", message, null)
            bytes += message.toByteArray().size
            // Keep the latest message intact, even when it alone exceeds the history budget.
            while (next.size > retainedMessages || (bytes > retainedBytes && next.size > 1)) {
                bytes -= next.removeAt(0).message.toByteArray().size
            }
        }
        matchID = poll.matchID; seatID = poll.seatID
        processedThrough = poll.revision; entries = next; messageBytes = bytes
    }

    companion object {
        private const val retainedMessages = 64
        private const val retainedBytes = 1024 * 1024
    }
}
